### PR TITLE
fix: duplicated Intellij IDE message when running flutter doctor

### LINKED ISSUE #98276: IntelliJ IDEA installed with JetBrains Toolbox duplicated in flutter doctor
JetBrains Toolbox is an official tool for managing JetBrains IDE installations and updates. On macOS, when you install an IDE from JetBrains, two apps will be created.

The location of one will not change, in the user's `Applications` folder. It's only 1.3MB and not the real IDE:

Screen Shot 2022-02-12 at 02 15 37

The location of the other, the real IDE, will change as the IDE is upgraded:

Screen Shot 2022-02-12 at 02 16 15

So when I run `flutter doctor -v` I will see two IntelliJ IDEA, one of which is fake:

Screen Shot 2022-02-12 at 02 21 08

### System Info

```
macOS 12.2.1
```

### Flutter Version

```
$ flutter --version
Flutter 2.10.1 • channel stable • https://github.com/flutter/flutter.git
Framework • revision db747aa133 (2 days ago) • 2022-02-09 13:57:35 -0600
Engine • revision ab46186b24
Tools • Dart 2.16.1 • DevTools 2.9.2
```

## Steps to Reproduce

1. Install JetBrains Toolbox
2. Install IntelliJ IDEA in JetBrains Toolbox
3. Run `flutter doctor` in Terminal

## Logs

<details>
<summary>
    <code>flutter doctor -v</code> output
</summary>

```
$ flutter doctor -v
[✓] Flutter (Channel stable, 2.10.1, on macOS 12.2.1 21D62 darwin-arm, locale en-US)
    • Flutter version 2.10.1 at /Users/reeky/.flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision db747aa133 (2 days ago), 2022-02-09 13:57:35 -0600
    • Engine revision ab46186b24
    • Dart version 2.16.1
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/reeky/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/reeky/Library/Android/sdk
    • Java binary at: /Users/reeky/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/211.7628.21.2111.8139111/Android
      Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Users/reeky/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/211.7628.21.2111.8139111/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Users/reeky/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Users/reeky/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/213.6777.52/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.64.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.34.0

[✓] Proxy Configuration
    • HTTP_PROXY is set
    • NO_PROXY is localhost,127.0.0.1,::1
    • NO_PROXY contains localhost
    • NO_PROXY contains 127.0.0.1
    • NO_PROXY contains ::1

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 12.2.1 21D62 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 97.0.4692.99

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by cychiang**
This PR fixes the duplicated message from `flutter doctor` when install `Intellij IDE` from `JetBrains Toolbox`. 

The solution is based on the #98276. Add a function to skip the creation of the validator for `Mac` when the key word `JetBrainsToolboxApp` is in the `info.plist`.

Before: 
Screenshot 2023-06-16 at 21 04 43

After: 
Screenshot 2023-06-16 at 21 13 15

fix #98276 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/intellij/intellij_validator.dart [hunk 1]
```diff
@@ -393,21 +393,29 @@ class IntelliJValidatorOnMac extends IntelliJValidator {
         fileSystem.path.join(homeDirPath, 'Applications'),
     ];
 
+    bool checkForJetBrainsToolboxWrapper(String installPath) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/intellij/intellij_validator.dart:396]: Some thoughts:

1. it seems like this will only work on macOS. Is this issue affecting Windows and Linux users also?
2. From reading [the issue](https://github.com/flutter/flutter/issues/98276), it sounds like the reason we have duplicate entries is because JetBrains toolbox generates one small executable that points to the actual install. Should we not then be skipping the small executable during detection? Or would there be cases when that would lead to zero IntelliJ's being detected?
3. Does this also affect Android Studio?

**Review by andrewkolos — CHANGES_REQUESTED** [review]
To expand on thought 2 from [@christopherfujino's earlier comment](https://github.com/flutter/flutter/pull/129030/files#r1239004710), it's possible that we find the same installation once [during our search in the /Applications directory](https://github.com/flutter/flutter/blob/b00f1c45998e19531ad2200c8b0925547363f4cf/packages/flutter_tools/lib/src/intellij/intellij_validator.dart#L391) and then again in our [Spotlight search](https://github.com/flutter/flutter/blob/b00f1c45998e19531ad2200c8b0925547363f4cf/packages/flutter_tools/lib/src/intellij/intellij_validator.dart#L432). I have not verified this.

The solution provided in this PR might be sufficient, but I think it might be more maintainable to instead deduplicate validators (i.e. those that have the same `installPath`) before returning the list rather than making an explicit check for toolbox. Let me know if you feel differently, @cychiang. This PR also needs test(s).

File: packages/flutter_tools/lib/src/intellij/intellij_validator.dart (hunk 1 above)
Comment (cychiang) [reply packages/flutter_tools/lib/src/intellij/intellij_validator.dart:396]: 1. Yes, it works on macOS but won't affect Windows and Linux because it isolates in a [IntelliJValidatorOnMac](https://github.com/flutter/flutter/blob/5e8900193dd097614075d83070740e78b92f687c/packages/flutter_tools/lib/src/intellij/intellij_validator.dart#L361) class specifically for [macOS](https://github.com/flutter/flutter/blob/5e8900193dd097614075d83070740e78b92f687c/packages/flutter_tools/lib/src/intellij/intellij_validator.dart#L73) only. There are separate implementations for Window and Linux. 
2. Yes, when I was working on the issue, I found that the validator is first looking for all possible candidate path that contains Intellij IDE, then extract and validate information from `plist` file. Not sure about the zero Intellij scenario, I will try to verify it. 
3. Android Studio is in another plugin [validator](https://github.com/flutter/flutter/blob/5e8900193dd097614075d83070740e78b92f687c/packages/flutter_tools/lib/src/android/android_studio_validator.dart#L18) so no affect on it.

**Comment by cychiang** [comment]
> Spotlight search 

I'm not fully sure on this part, but I don't see duplicate results from the Spotlight search. I will try to verify it and see why it doesn't happened. 

> but I think it might be more maintainable to instead deduplicate validators (i.e. those that have the same `installPath`) before returning the list rather than making an explicit check for toolbox.

I will try to see if there is another solution is more maintainable. But I would like to keep the change small and minimal. Will do the test after that as well.

**Comment by cychiang** [comment]
@andrewkolos @christopherfujino 
I want to give an update on the PR: 
- Replace `tool box` check with a step to remove duplicate validator based on its title before returning `List<DoctorValidator>`
- Add a test with multiple validators 
- Add missing mapping for `IntelliJ IDEA Community Edition.app` on [L375](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/intellij/intellij_validator.dart#L375)

## Root cause analysis: 

On [L416](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/intellij/intellij_validator.dart#L416) is trying to find candidates under certain paths: `Applications` and `${home_dir}/Applications` here is the definition on [L390](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/intellij/intellij_validator.dart#L390). 

The default folder to store all installed applications from JetBrains Toolbox is `${home_dir}/Library/Application Support/JetBrains/Toolbox`, and the location of the folder is not in the search paths: `Applications` and `${home_dir}/Applications`. 

The `JetBrains Toolbox` is installed in the `${home_dir}/Applications` folder by default, and it creates shortcut apps to the actual `IntelliJ App` for any installed app. The shortcut contains a `plist` file. The value of the `CFBundleIdentifier` from the file looks like this `com.jetbrains.toolbox.linkapp.*`, and it's not the target pattern(`com.jetbrains.intellij.ce` or `com.jetbrains.intellij*`) from the spotlight search. 

To sum up: 

- `Applications` and `${home_dir}/Applications` paths will result short cut apps
- The actual apps are in `${home_dir}/Library/Application Support/JetBrains/Toolbox` and those paths are not the target path from the `Directory` search
- Spotlight search looks for a keyword in `CFBundleIdentifier` with `mdfind` in case the JetBrains App installed outside of `Applications` and `${home_dir}/Applications` and the plist file in the shortcut app doesn't have the desired value

We don't see the similar behaviour in `IntelliJ IDEA Community Edition` from the latest JetBrains Toolbox because of [L375](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/intellij/intellij_validator.dart#L375) so it's not visible in the paths searched. So I included this change in my PR to have the same behaviour for all `IntelliJ IDEA` and remove duplicate validator before returning it.

File: packages/flutter_tools/lib/src/intellij/intellij_validator.dart (hunk 1 above)
Comment (cychiang) [reply packages/flutter_tools/lib/src/intellij/intellij_validator.dart]: To answer 2. if there is no validator then it won't show anything form `doctor`.

File: packages/flutter_tools/lib/src/intellij/intellij_validator.dart [hunk 2]
```diff
@@ -482,6 +483,13 @@ class IntelliJValidatorOnMac extends IntelliJValidator {
           ]),
       ));
     }
+
+    // Remove duplicate validator based on its title
+    final Set<String> validatorTitles = <String>{};
+    validators.removeWhere((DoctorValidator validator) {
+      return !validatorTitles.add(validator.title);
+    });
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/intellij/intellij_validator.dart:491]: Is it possible that a user's machine can have two installs with the same title (same version) that are both valid? Even if it's not possible now, it could be in the future.

Why not filter using install path instead?

Comment (cychiang) [reply packages/flutter_tools/lib/src/intellij/intellij_validator.dart:491]: Hi! I was trying to find another time to open another issue for what I will mention because it turns out like a rabbit hole and more issues from this. But let me try to answer your concern first: 

- why not filter using the install path instead: It creates another constraint to the install path since the JetBrains Toolbox supports changing the install path. For example, if we update the location to store the installed app and it contains a keyword that we want to filter, we might remove it accidentally. 
- If there are multiple installs with the same title(either `IntelliJ IDEA Ultimate Edition` or `IntelliJ IDEA Community Edition`) with the same install path and version: I would say it's a good scenario to remove it because they all provide the same information for `doctor`. But it needs extra care to handle possible multi-version and so on. 

I see this validator in another way. It's a validator to find all correct content and reduce it based on the version, install path and plugin path. Ideally, if we can parse the right install path from the shortcut app and update it, we can expect the same result from path-based and spotlight searches. It opens up the possibility of detecting multiple app versions and its plugins version. 

However, I further dug into this issue and discovered that we need to update the logic to find the correct install and plugin paths. The plugin path is also not working, and the issue I would like to open is related to this one. 

The result from the path search(toolbox) can get the correct plugin version, but the spotlight one is not. It's because of the [L510](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/intellij/intellij_validator.dart#L510) on `intellij_validator`. The shortcut app contains the `JetBrainsToolboxApp` key, and the value is the actual location of the app, the same as the spotlight search. It can form a correct path to the plugin's path. 

However, the result from the spotlight search depends on the keyword in the `plist`, and that keyword doesn't exist in the shortcut app. It helps to differentiate the result from path-based and spotlight searches. But the logic to find the correct plugin path when `JetBrainsToolboxApp` doesn't exist in the `plist` is outdated, resulting in an incorrect plugin path. 

To sum up: 
We might want to improve the following things: 

- Install and plugin path handling
- Reduce the result based on the candidate validator's title and version or other information before returning the list 

The issue is growing too much compared to the original possible solution, so I would like to work on those things separately to keep the changes minimal. 

It might be too long to read, and I would like your input @andrewkolos to continue the issue 😃

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/intellij/intellij_validator.dart:491]: Thanks for the write-up! I wouldn't be surprised that the logic around plugins doesn't work properly. I've worked on the `AndroidStudioValidator` recently, and I am pretty sure that class has similar issues with detecting the Dart and Flutter plugins. I agree that this is a separate issue, so please feel free to file that with the `tool` label.

Moving back to this issue/PR, I realize I was misremembering the issue that this is trying to fix. Indeed, we can't use the install path to filter here. In that case, should we go back to using the plist file or looking for `Jetbrains Toolbox` in the installation path? I think filtering on the title works, but the result could change based on the order in which we detect installations of IntelliJ.

**Review by andrewkolos — CHANGES_REQUESTED** [review]
See https://github.com/flutter/flutter/pull/129030/files#r1244252129

**Comment by cychiang** [comment]
Hi @andrewkolos I updated the code and use `CFBundleIdentifier` to filter `JetBrains Toolbox` results.

Here is the short summary: 

- Filter `JetBrains Toolbox` based on `CFBundleIdentifier` in plist file, and I think it's a reliable way to validate an app is a shortcut or a executable file. 
- When adding spotlight results, there is a check to validate the `installPath` to avoid multiple validator with the same `installPath`. So there is no need to consider this scenario.

File: packages/flutter_tools/lib/src/intellij/intellij_validator.dart [hunk 3]
```diff
@@ -482,6 +483,16 @@ class IntelliJValidatorOnMac extends IntelliJValidator {
           ]),
       ));
     }
+
+    // Remove JetBrains Toolbox results
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/intellij/intellij_validator.dart:487]: Consider explaining _why_ we are removing these results. This is more important than explaining _what_ we are removing.

```suggestion
    // Remove JetBrains Toolbox link apps. These tiny apps just
    // link to the full app, will get detected elsewhere in our search.
```

**Review by andrewkolos — CHANGES_REQUESTED** [review]
Sorry for taking a while to get back to this PR. I have just one nitpick. Otherwise, LGTM.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Review by andrewkolos — APPROVED** [review]
LGTM
